### PR TITLE
Rework withdraw liquidity

### PR DESCRIPTION
**Pull request by amityadav0**
- Rework withdraw deposit similarly to what has been done with add_deposit

### CODE REVIEW

File: nearswap/src/constants.rs [hunk 1]
```diff
@@ -1,7 +1,9 @@
-use near_sdk::{env, Balance};
-
 // pub const STORAGE_PRICE_PER_BYTE: Balance = env::STORAGE_PRICE_PER_BYTE;
 
 /// Minimum Account Storage used for account registration.
 /// 64 (AccountID bytes) + 2*8 (int32) + byte
 pub const INIT_ACCOUNT_STORAGE: u64 = 64 + 16 + 4;
+
+pub const NDENOM: u128 = 1_000_000_000_000_000_000_000_000;
```
Comment (robert-zaremba) [inline nearswap/src/constants.rs:7]:
```suggestion
/// 1 NEAR in yocto = 1e24
pub const NDENOM: u128 = 1_000_000_000_000_000_000_000_000;
```

File: nearswap/src/lib.rs [hunk 2]
```diff
@@ -156,59 +157,50 @@ impl NearSwap {
...
-    /// exchengable between different pools.
+    /// `min_ynear` of Near and `min_tokens` of tokens. Shares are not
+    /// exchagable between different pools.
     pub fn withdraw_liquidity(
         &mut self,
         token: AccountId,
-        shares: U128,
         min_ynear: U128,
         min_tokens: U128,
+        shares: U128,
```
Comment (robert-zaremba) [inline nearswap/src/lib.rs:167]: why reordering?

File: nearswap/src/lib.rs [hunk 3]
```diff
@@ -696,7 +688,17 @@ mod tests {
         let ynear_deposit_with_storage = ynear_deposit + NEP21_STORAGE_DEPOSIT;
         ctx.set_deposit(ynear_deposit_with_storage);
 
-        c.add_liquidity(t.clone(), token_deposit.into(), ynear_deposit.into());
+        let account_deposit = AccountDeposit {
+            ynear: 2*ynear_deposit + NDENOM,
+            storage_used: 10,
+            tokens: [(t.clone(), token_deposit * 11)]
+                .iter()
+                .cloned()
```
Comment (robert-zaremba) [inline nearswap/src/lib.rs:696]: do we need `cloned` here?

File: nearswap/src/lib.rs [hunk 4]
```diff
@@ -696,7 +688,17 @@ mod tests {
...
+                .collect(),
+        };
+        c.set_deposit(&a.clone(), &account_deposit);
+
+        c.add_liquidity(t.clone(), ynear_deposit.into(), token_deposit.into(), U128(0));
```
Comment (robert-zaremba) [inline nearswap/src/lib.rs:701]:
```suggestion
        c.set_deposit(&a.clone(), &account_deposit);
        c.add_liquidity(t.clone(), ynear_deposit.into(), token_deposit.into(), U128(0));
```

File: nearswap/src/pool.rs [hunk 5]
```diff
@@ -128,4 +128,36 @@ impl Pool {
         }
         return (ynear, added_tokens, shares_minted);
     }
+
+    /// Redeems `shares` for liquidity stored in this pool with condition of getting at least
+    /// `min_ynear` of Near and `min_tokens` of tokens. Shares are not
+    /// exchagable between different pools.
+    /// withdraw liquidity from pool and deposit to account deposit
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:135]:
```suggestion
    /// Withdraw `shares` for liquidity stored in this pool and transfer them to the caller deposit account. User can require 
    /// getting at least `min_ynear` of Near and `min_tokens` of tokens. The function panic if the condition is not met. 
    /// Shares are not exchangeable between different pools.
```

File: nearswap/src/pool.rs [hunk 6]
```diff
@@ -128,4 +128,36 @@ impl Pool {
...
+                ynear, token_amount
+            )
+        );
+    
+        self.shares.insert(caller, &(current_shares - shares));
+        self.total_shares -= shares;
+        self.tokens -= token_amount;
+        self.ynear -= ynear;
+
+        return (shares, ynear, token_amount);
```
Comment (robert-zaremba) [inline nearswap/src/pool.rs:161]: why do we return `shares`? It's not modified.

File: nearswap/src/lib.rs [hunk 7]
```diff
@@ -783,6 +795,82 @@ mod tests {
...
+        // total supply of an unknown token must be 0
+        assert_eq!(
+            to_num(c.total_supply("unknown-token".to_string())),
+            0,
+            "total supply of other token shouldn't change"
+        );
+
+        println!(">> adding liquidity - second time with minted shares");
+
+        let min_shares = 30000000000000000000000000;
```
Comment (robert-zaremba) [inline nearswap/src/lib.rs:852]: please use / create function `fn to_yocto(amount &str) -> u128)`

Comment (robert-zaremba) [reply nearswap/src/lib.rs:852]: example usage: `to_yocto("0.2") -> 200e21 = 0.2e24`

Comment (robert-zaremba) [reply nearswap/src/lib.rs:852]: There might be already a function for this somewhere.

**Review by robert-zaremba — APPROVED** [review]
pre approving.

File: nearswap/src/lib.rs (hunk 2 above)
Comment (amityadav0) [reply nearswap/src/lib.rs:167]: was trying to keep order same as add_liquidity. will keep same it not required

File: nearswap/src/lib.rs (hunk 3 above)
Comment (amityadav0) [reply nearswap/src/lib.rs:696]: yes, without clone it is not working

File: nearswap/src/lib.rs (hunk 7 above)
Comment (amityadav0) [reply nearswap/src/lib.rs:852]: ok

File: nearswap/src/pool.rs (hunk 6 above)
Comment (amityadav0) [reply nearswap/src/pool.rs:161]: removing

File: nearswap/src/lib.rs (hunk 7 above)
Comment (amityadav0) [reply nearswap/src/lib.rs:851]: it is in near_sdk_sim

**Pull request merged**